### PR TITLE
Update pybind11 to latest.

### PR DESCRIPTION
**Pull request by ScottTodd**
This fixes a build failure of the Python bindings on Windows. We were testing Python 3.10.8, but now we are testing Python 3.11.0.
* Old 3.10.8 success: https://github.com/iree-org/iree/actions/runs/3452338984/jobs/5762178428
* Old 3.11.0 failure: https://github.com/iree-org/iree/actions/runs/3461697554/jobs/5779634638

New successful run on Python 3.11.0: https://github.com/iree-org/iree/actions/runs/3463493036/jobs/5783809576

### CODE REVIEW

**Comment by GMNGeoffrey** [comment]
Do we need to raise the version of pybind in our pip installs also? https://github.com/iree-org/iree/blob/03190c689ae29bca4bd5ea71d5793811c47a90a2/runtime/bindings/python/iree/runtime/build_requirements.txt#L7

**Comment by ScottTodd** [comment]
> Do we need to raise the version of pybind in our pip installs also?
> 
> https://github.com/iree-org/iree/blob/03190c689ae29bca4bd5ea71d5793811c47a90a2/runtime/bindings/python/iree/runtime/build_requirements.txt#L7

Only if using Python 3.11.0 (on Windows?). Looks like 2.8.0 from that file was released Oct 4, 2021: https://pypi.org/project/pybind11/#history. I'm not sure what our policy is / should be for dependencies in `bindings/python/iree/runtime/build_requirements.txt` - should we include minimum versions for some configuration, or bump them regularly?

**Comment by ScottTodd** [comment]
Okay, bumped a few other files to pybin11 2.10.1

https://pybind11.readthedocs.io/en/latest/changelog.html
> Version 2.10.1 (Oct 31, 2022)
> This is the first version to fully support embedding the newly released Python 3.11.

**Comment by GMNGeoffrey** [comment]
Yeah I don't know :thinking: I think generally we should try to be more permissive in cases where it's harder for people to update. If someone depends on another project that needs pybind, it might be annoying to force them onto a later version. I think the pip package must server a slightly different purpose than the source dependency. Do they have a canonical correspondence?

**Comment by GMNGeoffrey** [comment]
Ok, yeah it looks like there is a direct correspondence. Can we use a commit from the stable branch or their latest release instead of just the head commit? I think we can also configure our .gitmodules file to indicate that for pybind11 we track the stable branch (though I'm not sure how much we ever use submodules' magic updating features)

**Comment by GMNGeoffrey** [comment]
[sorry for triple-posting]

> Only if using Python 3.11.0 (on Windows?)

I think you can actually express these constraints with pip requirements specifiers: https://pip.pypa.io/en/stable/reference/requirement-specifiers/

We can't really do that with our source dependency though, so probably not a good idea.

File: runtime/bindings/python/iree/runtime/build_requirements.txt [hunk 1]
```diff
@@ -4,7 +4,7 @@
 #   python -m pip install -r runtime/bindings/python/iree/runtime/build_requirements.txt
 
 numpy>=1.19.4
-pybind11>=2.8.0
+pybind11>=2.10.1
```
Comment (GMNGeoffrey) [inline runtime/bindings/python/iree/runtime/build_requirements.txt:7]: So.... this now requires rebuilding the docker containers... Sorry :grimacing:

Comment (ScottTodd) [reply runtime/bindings/python/iree/runtime/build_requirements.txt:7]: Argh, I searched in docker/ and saw no matches. Okay, I can address your comments tomorrow then. `build_test_all_windows` will remain broken on postsubmit until then >_>

Comment (GMNGeoffrey) [reply runtime/bindings/python/iree/runtime/build_requirements.txt:7]: Could you just pin the python version there? I think we actually should do that in general. Things that run pointing at a specific commit shouldn't start failing for unrelated reasons

Comment (ScottTodd) [reply runtime/bindings/python/iree/runtime/build_requirements.txt:7]: Yeah, that would be good to do, though if we were pinned to 3.10 we wouldn't have noticed this break :P

Comment (GMNGeoffrey) [reply runtime/bindings/python/iree/runtime/build_requirements.txt:7]: I think if we were pinned to 3.10 we would've noticed that that wasn't the latest Python version we supported and we should actually be pinned to 3.11 :-) IMO we should try to record min/max versions of what we support and then pin those. As new versions come out, we should roll forward the max version

Comment (ScottTodd) [reply runtime/bindings/python/iree/runtime/build_requirements.txt:7]: Do we officially support 3.11? I think we're only publishing wheels for 3.7, 3.8, 3.9, and 3.10? https://github.com/iree-org/iree/blob/acb2688c06ab8c0907eaa5462b9d2ede15be9cd1/build_tools/python_deploy/build_linux_packages.sh#L46

Comment (GMNGeoffrey) [reply runtime/bindings/python/iree/runtime/build_requirements.txt:7]: Oh well all the more reason to pin to 3.10 then 😛 I'm fine with us having builds that try to test us against some moving head, but they shouldn't be the same builds that try to tell us the health of a specific commit. The latter shouldn't have a time dependence

**Comment by ScottTodd** [comment]
Closing this in favor of https://github.com/iree-org/iree/pull/11175. We can update pybind11 to >=2.10.1 when we next rebuild our Docker images.

**Pull request closed without merging**